server/datastore/mysql/vpp.go

```diff
@@ -831,28 +1014,73 @@ func (ds *Datastore) GetVPPTokenByTeamID(ctx context.Context, teamID *uint) (*fl
 		}
 	}
 
-	tokDec, err := decrypt([]byte(vppTok.Token), ds.serverPrivateKey)
+	tokDec, err := decrypt([]byte(tokEnc.Token), ds.serverPrivateKey)
 	if err != nil {
 		return nil, ctxerr.Wrap(ctx, err, "decrypting vpp token with serverPrivateKey")
 	}
-	vppTok.Token = string(tokDec)
 
-	return &vppTok, nil
+	if err := sqlx.SelectContext(ctx, ds.reader(ctx), &tokTeams, stmtTeamNames, tokEnc.ID); err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "retrieving vpp token team information")
+	}
+
+	tok := &fleet.VPPTokenDB{
+		ID:        tokEnc.ID,
+		OrgName:   tokEnc.OrgName,
+		Location:  tokEnc.Location,
+		RenewDate: tokEnc.RenewDate,
+		Token:     string(tokDec),
+	}
+
+	if tokTeams == nil {
+		// Not assigned, no need to loop over teams
+		return tok, nil
+	}
+
+TEAMLOOP:
+	for _, team := range tokTeams {
+		switch team.NullTeam {
+		case fleet.NullTeamAllTeams:
+			// This should only be possible if there are no other teams
+			// Make sure something array is non-nil
+			if len(tokTeams) != 1 {
+				return nil, ctxerr.Errorf(ctx, "team \"%s\" belongs to All teams, and %d other team(s)", tok.OrgName, len(tokTeams)-1)
+			}
+			tok.Teams = []fleet.TeamTuple{}
+			break TEAMLOOP
+		case fleet.NullTeamNoTeam:
+			tok.Teams = append(tok.Teams, fleet.TeamTuple{
+				ID:   0,
+				Name: fleet.TeamNameNoTeam,
+			})
+		case fleet.NullTeamNone:
+			// Regular team
+			tok.Teams = append(tok.Teams, fleet.TeamTuple{
+				ID:   *team.TeamID,
+				Name: team.Name,
+			})
+		}
+	}
+
+	return tok, nil
 }
 
-func (ds *Datastore) checkVPPNullTeam(ctx context.Context, currentID *uint, nullTeam fleet.NullTeamType) error {
-	nullTeamStmt := `SELECT id FROM vpp_tokens WHERE null_team_type = ?`
+func checkVPPNullTeam(ctx context.Context, tx sqlx.ExtContext, currentID *uint, nullTeam fleet.NullTeamType) error {
+	nullTeamStmt := `SELECT vpp_token_id FROM vpp_token_teams WHERE null_team_type = ?`
 
 	if nullTeam != fleet.NullTeamNone {
 		var id uint
-		if err := sqlx.GetContext(ctx, ds.reader(ctx), &id, nullTeamStmt, nullTeam); err != nil {
-			if !errors.Is(err, sql.ErrNoRows) {
-				return ctxerr.Wrap(ctx, err, "checking for nullteam constraints")
-			}
-		} else {
-			if currentID == nil || *currentID != id {
-				return ctxerr.Errorf(ctx, "vpp token for team %s already exists", nullTeam)
+		row := tx.QueryRowxContext(ctx, nullTeamStmt, nullTeam)
```

Comment:
It's a bit nicer to use `sqlx.GetContext(...)` for this case, instead of getting the row and scanning in a destination.